stdlib/Distributed/src/remotecall.jl

```diff
@@ -548,11 +554,28 @@ Further calls to `fetch` on the same reference return the cached value. If the r
 is an exception, throws a [`RemoteException`](@ref) which captures the remote exception and backtrace.
 """
 function fetch(r::Future)
-    r.v !== nothing && return something(r.v)
-    v = call_on_owner(fetch_ref, r)
-    r.v = Some(v)
+    v_cache = @atomic :acquire r.v
+    v_cache !== nothing && return something(v_cache)
+
```

Comment:
If I put the whole thing in a `lock(r.lock)` then a local `put!` will never be able to enter the lock, because the `fetch` will be stuck on the `fetch(rv.c)`. And in `put!` if i put into the channel outside of the lock I don't have control over who caches the local value first

I'm not sure what else could I improve here.
The local cache is always set once. The paths from `fetch(rv.c)` and `call_on_owner` will rightfully try to cache the value and only fail if it was already cached or `put!` was locally.

Reply:
Okay, yeah, I see how it is complicated by the fact that we might be the owner, so we are waiting for someone else to set the value before we can return it here. But it seems like if this fails in the remote case, it is because there was a thread-synchronization error, which could happen because this lock was not being held across here.

Reply:
the call on owner case won't fail due to this lock, because it fetches on a remote channel

this `fetch` function isn't used in the `call_on_owner` scenario

This looks safe to me as it is right now because of that, but I haven't dwelled that deep into the remote scenarios